### PR TITLE
Update cookiecutter

### PR DESCRIPTION
**Pull request by Zeitsperre**
## Overview

Changes:

* Updated the cookiecutter to the latest commit
* Updated pre-commit to use newer versions and coding conventions
* Updated the code base with pre-commit
* Synchronizes dependencies between conda and PyPI

## Related Issue / Discussion

There's a Pull Request open at https://github.com/bird-house/cookiecutter-birdhouse/pull/120 (waiting on review) to bring the original recipe up to speed. Python3.9 support is already nearing End-of-Life.

## Additional Information

This fast-forward is still very behind. This is already a major improvement, though it seems as though the recipe has a few regressions in it; I've done my best to integrate the two.

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@cehbrecht 

A few issues:
- I think the changes at Copernicus have rendered most of the smoke tests unable to pass. Either that or there's a configuration missing.
- The docs are failing to build the API due to issues with dependencies. This was already seen in the "latest" builds of the documentation. Not sure how long that has been the case.
- I think in order to address some of these problems, we need new releases of `clisops` and `daops`; This is especially true if we want Python3.12 support.

**Comment by cehbrecht** [comment]
> @cehbrecht
> 
> A few issues:
> 
> * I think the changes at Copernicus have rendered most of the smoke tests unable to pass. Either that or there's a configuration missing.
> * The docs are failing to build the API due to issues with dependencies. This was already seen in the "latest" builds of the documentation. Not sure how long that has been the case.
> * I think in order to address some of these problems, we need new releases of `clisops` and `daops`; This is especially true if we want Python3.12 support.

@Zeitsperre the smoke tests only work on a deployed Rook instance with data mounted. I use them to make sure everything is ok *after* the rook installation.

**Comment by Zeitsperre** [comment]
@cehbrecht 

I managed to figure out why the docs were failing: It was because almost every library was being mocked in the sphinx configuration, except for `clisops`, and since that library relies on the `roocs_utils` configuration-fetching code, this was not being imported properly and causing really bizarre build errors.

If things are looking good here, perhaps we can set ReadTheDocs to build Pull Requests and merge this and #259 ? These will both help with achieving #252.

**Comment by Zeitsperre** [comment]
@tlvu Let me know if the Docker configuration is good for you!

**Comment by Zeitsperre** [comment]
@tlvu and @cehbrecht,

As far as I'm concerned, this is about as modern as I can make `rook` before needing to delve into the code base. Waiting on review and approval before this gets merged into #259 and a new version can be released and placed on Docker.

I haven't implemented any major coding convention/style changes yet, but when that happens (upstream), changes will be pushed to all services based on the template.

Waiting on approval before starting the update process.

**Comment by tlvu** [comment]
@Zeitsperre sorry for the review delay.  I've been trying get Rocky9 working for our Jenkins and I think I am very close.  Should be able to get to this PR by the end of this week.

**Comment by Zeitsperre** [comment]
@tlvu Hoping to wrap this up. Let me know if anything's missing!

File: .github/workflows/main.yml
Comment (tlvu) [inline .github/workflows/main.yml:55]: Not working with py 3.12?

File: .readthedocs.yml
Comment (tlvu) [inline .readthedocs.yml:13]: nice !

File: LICENSE
Comment (tlvu) [inline LICENSE:190]: Should this be set to a real year and owner name?

File: Makefile
Comment (tlvu) [inline Makefile:152]: Haha, old memories of Travis-CI

**Review by tlvu — APPROVED** [review]
Looks good to me.  Apologies I got side-tracked !

Just some very minor feedback.

File: LICENSE
Comment (Zeitsperre) [reply LICENSE:190]: No, this is the boilerplate license file that needs to be in the files.

In order to apply the license, the block of text itself needs to be added into the source files, with the changes for year and authorship. Learned this not too long ago.

There are some more issues around using the Apache v2.0 License as well; I might bring this up in our next team meeting.

File: Makefile
Comment (Zeitsperre) [reply Makefile:152]: RIP

File: .github/workflows/main.yml
Comment (Zeitsperre) [reply .github/workflows/main.yml:55]: There's a problem with `clisops` that prevents us from being able to install it in Python3.12. The last `xarray` version that rook supports is very old due to a regression that impacts `cftime` time encodings.

We would need to open this pin and deal with the problem somehow. Until then, my hands are tied.

**Pull request merged**
